### ISSUE TITLE
SFT loss not converging

### ISSUE THREAD

**Issue by FlyingPiggyKing** [lvwerra/trl#491]
I wish to fine tune the model on lvwerra/stack-exchange-paired dataset, my base model is yahma/llama-7b-hf, which is a hf format llama 7b, I have 2 V100V3 card. The training can be launched without problem, the problem is the loss is not converging, even explode later. below is my launch command:

OMP_NUM_THREADS=4 WORLD_SIZE=2 CUDA_VISIBLE_DEVICES=0,1 torchrun --nnodes 1  --nproc_per_node 2 ./supervised_finetuning.py --model_path=yahma/llama-7b-hf --no_gradient_checkpointing --learning_rate 1e-5  --num_workers 8 --output_dir ./model_output

The loss is as below:

_trainable params: 8388608 || all params: 6746804224 || trainable%: 0.12433454005023165
Training...
***** Running training *****
  Num examples = 7,440,550
  Num Epochs = 1
  Instantaneous batch size per device = 4
  Total train batch size (w. parallel, distributed & accumulation) = 8
  Gradient Accumulation steps = 1
  Total optimization steps = 10,000
  Number of trainable parameters = 8,388,608

  0%|                                                 | 0/10000 [00:00<?, ?it/s]`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`...
`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`...
{'loss': 1.6223, 'learning_rate': 5e-06, 'epoch': 0.0}
{'loss': 1.6505, 'learning_rate': 1e-05, 'epoch': 0.0}
{'loss': 1.5876, 'learning_rate': 9.999370638369377e-06, 'epoch': 0.0}
{'loss': 1.5825, 'learning_rate': 9.997482711915926e-06, 'epoch': 0.0}
{'loss': 1.5144, 'learning_rate': 9.994336695915041e-06, 'epoch': 0.0}
{'loss': 1.5767, 'learning_rate': 9.989933382359423e-06, 'epoch': 0.0}
{'loss': 1.6146, 'learning_rate': 9.984273879759713e-06, 'epoch': 0.0}
{'loss': 1.5778, 'learning_rate': 9.977359612865424e-06, 'epoch': 0.0}
{'loss': 1.5852, 'learning_rate': 9.969192322306271e-06, 'epoch': 0.0}
{'loss': 1.5882, 'learning_rate': 9.959774064153977e-06, 'epoch': 0.0}
{'loss': 1.5557, 'learning_rate': 9.949107209404664e-06, 'epoch': 0.0}
{'loss': 1.5723, 'learning_rate': 9.937194443381972e-06, 'epoch': 0.0}
{'loss': 1.531, 'learning_rate': 9.924038765061042e-06, 'epoch': 0.0}
{'loss': 1.5471, 'learning_rate': 9.909643486313533e-06, 'epoch': 0.0}
{'loss': 1.5437, 'learning_rate': 9.894012231073895e-06, 'epoch': 0.0}
{'loss': 1.5678, 'learning_rate': 9.877148934427037e-06, 'epoch': 0.0}
{'loss': 1.6571, 'learning_rate': 9.859057841617709e-06, 'epoch': 0.0}
{'loss': 1.7075, 'learning_rate': 9.839743506981783e-06, 'epoch': 0.0}
{'loss': 2.3136, 'learning_rate': 9.819210792799711e-06, 'epoch': 0.0}
{'loss': 5.1728, 'learning_rate': 9.797464868072489e-06, 'epoch': 0.0}
{'loss': 0.0, 'learning_rate': 9.774511207220369e-06, 'epoch': 0.0}
{'loss': 0.0, 'learning_rate': 9.750355588704728e-06, 'epoch': 0.0}
{'loss': 0.0, 'learning_rate': 9.725004093573343e-06, 'epoch': 0.0}_

Is this normally? how should I prevent Loss explosion?

**Comment by lvwerra** [comment]
I would try to vary the LR, have you tried gowing smaller?

**Comment by FlyingPiggyKing** [comment]
I have tried increased the learning rate, I also tried increase the batch size, increase the gradient_accumulation_steps, I see you are using 8 cards, which means 32 samples in a batch, unfortuntely none of these works, I can't see loss converging.

I will try decrease the learning rate later.

**Comment by lvwerra** [comment]
Have you tried to **decrease** the learning rate?

**Comment by FlyingPiggyKing** [comment]
tring with learning rate 1e-8, loss as below:
<img width="1074" alt="image" src="https://github.com/lvwerra/trl/assets/7313383/030219fd-be1f-4c86-bed1-b2f8a30a4c62">
{'loss': 1.621, 'learning_rate': 2.113183355555904e-09, 'epoch': 0.02}          
{'loss': 1.5851, 'learning_rate': 2.08706867744802e-09, 'epoch': 0.02}          
{'loss': 1.6364, 'learning_rate': 2.0610737385376347e-09, 'epoch': 0.02}        
{'loss': 1.5912, 'learning_rate': 2.0351996073748715e-09, 'epoch': 0.02}        
{'loss': 1.5526, 'learning_rate': 2.00944734754392e-09, 'epoch': 0.02}          
{'loss': 1.5678, 'learning_rate': 1.9838180176193177e-09, 'epoch': 0.02}        
{'loss': 1.5841, 'learning_rate': 1.958312671122434e-09, 'epoch': 0.02}         
{'loss': 1.6254, 'learning_rate': 1.932932356478168e-09, 'epoch': 0.02}         
{'loss': 1.5817, 'learning_rate': 1.907678116971843e-09, 'epoch': 0.02}         
{'loss': 1.5979, 'learning_rate': 1.8825509907063328e-09, 'epoch': 0.02}        
{'loss': 1.5794, 'learning_rate': 1.8575520105593818e-09, 'epoch': 0.02}        
{'loss': 1.5769, 'learning_rate': 1.8326822041411524e-09, 'epoch': 0.02}        
{'loss': 1.5517, 'learning_rate': 1.8079425937519728e-09, 'epoch': 0.02}        
{'loss': 1.5505, 'learning_rate': 1.783334196340331e-09, 'epoch': 0.02}         
{'loss': 1.5473, 'learning_rate': 1.758858023461059e-09, 'epoch': 0.02}         
{'loss': 1.5242, 'learning_rate': 1.7345150812337561e-09, 'epoch': 0.02}        
{'loss': 1.5715, 'learning_rate': 1.710306370301437e-09, 'epoch': 0.02}         
{'loss': 1.6583, 'learning_rate': 1.6862328857893855e-09, 'epoch': 0.02}        
{'loss': 1.6113, 'learning_rate': 1.66229561726426e-09, 'epoch': 0.02}          
{'loss': 1.6116, 'learning_rate': 1.6384955486934156e-09, 'epoch': 0.02}        
{'loss': 1.6263, 'learning_rate': 1.6148336584044537e-09, 'epoch': 0.02}        
{'loss': 1.6413, 'learning_rate': 1.5913109190450032e-09, 'epoch': 0.02}        
{'loss': 1.6021, 'learning_rate': 1.567928297542749e-09, 'epoch': 0.02}         
 75%|█████████████████████████▍        | 3733/5000 [18:44:50<7:09:22, 20.33s/it]

it's alomost 4000 steps, each step 32 examples. Good news, not exploded, bad news, no obverious converging.

I'm going to try 1e-6

**Comment by FlyingPiggyKing** [comment]
the result of lr 1e-6
<img width="1073" alt="image" src="https://github.com/lvwerra/trl/assets/7313383/d242da36-46a5-44e6-9429-33bacbbd14d9">

obviously 1e-6 is better, I can see slowly converging

**Comment by FlyingPiggyKing** [comment]
@lvwerra , from you perspective, do you think this loss converging is normal? could you share you loss number? like how much steps have you trained, and what train loss and eval loss have you got.

**Comment by lvwerra** [comment]
This depends a lot on your data and manually inspecting the model by looking at some generations after the training might give you some intuition if it worked.

**Comment by FlyingPiggyKing** [comment]
> This depends a lot on your data and manually inspecting the model by looking at some generations after the training might give you some intuition if it worked.

do you mean the loss here is already very small, whether converging or not is not a good sign to indicate the training process  is moving forward to a correct direction, instead to check the model output to see whether it behave as a StackExchange man is a better way?

**Comment by lvwerra** [comment]
Converging is a good sign but if you want to figure out if it converged enough looking at some generations might help. Also doing a more extensive LR sweep might help find an even better setting.
